src/Assistant.spec.ts

```diff
@@ -0,0 +1,445 @@
+import 'mocha';
+import { assert } from 'chai';
+import sinon from 'sinon';
+import rewiremock from 'rewiremock';
+import { WebClient } from '@slack/web-api';
+import {
+  Assistant,
+  AssistantMiddlewareArgs,
+  AllAssistantMiddlewareArgs,
+  AssistantMiddleware,
+  AssistantConfig,
+  AssistantThreadStartedMiddlewareArgs,
+  AssistantThreadContextChangedMiddlewareArgs,
+  AssistantUserMessageMiddlewareArgs,
+} from './Assistant';
+import { Override } from './test-helpers';
+import { AllMiddlewareArgs, AnyMiddlewareArgs, AssistantThreadStartedEvent, Middleware, SlackEventMiddlewareArgs } from './types';
+import { AssistantInitializationError } from './errors';
+
+async function importAssistant(overrides: Override = {}): Promise<typeof import('./Assistant')> {
+  return rewiremock.module(() => import('./Assistant'), overrides);
+}
+
+const MOCK_FN = async () => { };
+
+const MOCK_CONFIG_SINGLE = {
+  threadStarted: MOCK_FN,
+  threadContextChanged: MOCK_FN,
+  userMessage: MOCK_FN,
+};
+
+const MOCK_CONFIG_MULTIPLE = {
+  threadStarted: [MOCK_FN, MOCK_FN],
+  threadContextChanged: [MOCK_FN],
+  userMessage: [MOCK_FN, MOCK_FN, MOCK_FN],
+};
+
+describe('Assistant class', () => {
+  describe('constructor', () => {
+    it('should accept config as single functions', async () => {
+      const assistant = new Assistant(MOCK_CONFIG_SINGLE);
+      assert.isNotNull(assistant);
+    });
+
+    it('should accept config as multiple functions', async () => {
+      const assistant = new Assistant(MOCK_CONFIG_MULTIPLE);
+      assert.isNotNull(assistant);
+    });
+
+    describe('validate', () => {
+      it('should throw an error if config is not an object', async () => {
+        const { validate } = await importAssistant();
+
+        // intentionally casting to AssistantConfig to trigger failure
+        const badConfig = '' as unknown as AssistantConfig;
+
+        const validationFn = () => validate(badConfig);
+        const expectedMsg = 'Assistant expects a configuration object as the argument';
+        assert.throws(validationFn, AssistantInitializationError, expectedMsg);
+      });
+
+      it('should throw an error if required keys are missing', async () => {
+        const { validate } = await importAssistant();
+
+        // intentionally casting to AssistantConfig to trigger failure
+        const badConfig = {
+          threadStarted: async () => { },
+        } as unknown as AssistantConfig;
+
+        const validationFn = () => validate(badConfig);
+        const expectedMsg = 'Assistant is missing required keys: threadContextChanged, userMessage';
```

Comment:
This test fails. Also, some code do not compile. To review this PR, I've modfiied the code this way: https://github.com/seratch/bolt-js/commit/0d4b52be2fdae38a2be47d09ba8646d6b6d76d35

Reply:
Yep, I have a TODO to update the tests, though I'm not sure why I don't experience the same blocking behavior you are. Thanks for providing your updates directly – will make use of them.